protocols/relay/src/v2/client/transport.rs

```diff
@@ -0,0 +1,405 @@
+use crate::v2::RequestId;
+use futures::channel::mpsc;
+use futures::channel::oneshot;
+use futures::future::{ready, BoxFuture, Future, FutureExt, Ready};
+use futures::sink::SinkExt;
+use futures::stream::{Stream, StreamExt};
+use libp2p_core::multiaddr::{Multiaddr, Protocol};
+use libp2p_core::transport::{ListenerEvent, TransportError};
+use libp2p_core::{PeerId, Transport};
+use std::collections::VecDeque;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use thiserror::Error;
+
+/// A [`Transport`] enabling client relay capabilities.
+///
+/// Needs to be combined with another protocol, e.g. via
+/// [`OrTransport`](libp2p_core::transport::choice::OrTransport).
+///
+/// Allows the local node to:
+///
+/// 1. Establish relayed connections by dialing `/p2p-circuit` addresses.
+///
+///    ```
+///    # use libp2p_core::{Multiaddr, multiaddr::{Protocol}, Transport, PeerId};
+///    # use libp2p_core::transport::memory::MemoryTransport;
+///    # use libp2p_core::transport::choice::OrTransport;
+///    # use libp2p_relay::v2::client;
+///    let actual_transport = MemoryTransport::default();
+///    let (relay_transport, behaviour) = client::Client::new_transport_and_behaviour(
+///        PeerId::random(),
+///    );
+///    let transport = OrTransport::new(relay_transport, actual_transport);
+///    # let relay_id = PeerId::random();
+///    # let destination_id = PeerId::random();
+///    let dst_addr_via_relay = Multiaddr::empty()
+///        .with(Protocol::Memory(40)) // Relay address.
+///        .with(Protocol::P2p(relay_id.into())) // Relay peer id.
+///        .with(Protocol::P2pCircuit) // Signal to connect via relay and not directly.
+///        .with(Protocol::P2p(destination_id.into())); // Destination peer id.
+///    transport.dial(dst_addr_via_relay).unwrap();
+///    ```
+///
+/// 3. Listen for incoming relayed connections via specific relay.
+///
+///    ```
+///    # use libp2p_core::{Multiaddr, multiaddr::{Protocol}, Transport, PeerId};
+///    # use libp2p_core::transport::memory::MemoryTransport;
+///    # use libp2p_core::transport::choice::OrTransport;
+///    # use libp2p_relay::v2::client;
+///    let actual_transport = MemoryTransport::default();
+///    let (relay_transport, behaviour) = client::Client::new_transport_and_behaviour(
+///        PeerId::random(),
+///    );
+///    let transport = OrTransport::new(relay_transport, actual_transport);
+///    let relay_addr = Multiaddr::empty()
+///        .with(Protocol::Memory(40)) // Relay address.
+///        .with(Protocol::P2p(PeerId::random().into())) // Relay peer id.
+///        .with(Protocol::P2pCircuit); // Signal to listen via remote relay node.
+///    transport.listen_on(relay_addr).unwrap();
+///    ```
+#[derive(Clone)]
+pub struct ClientTransport {
+    to_behaviour: mpsc::Sender<TransportToBehaviourMsg>,
+}
+
+impl ClientTransport {
+    /// Create a new [`ClientTransport`] by wrapping an existing [`Transport`] in a
+    /// [`ClientTransport`].
```

Comment:
```suggestion
    /// Create a new [`ClientTransport`]. 
    ///
    /// Note: The transport only handles listening and dialing on relayed [`Multiaddr`], and depends on an other transport to do the actual transmission of data. They should be combined through the [`OrTransport`].
```
Feel free to adjust my wording.

Reply:
Oh, thanks for catching the outdated comment. Fixed with 9a0f886637587910c2664d0f87ecc6897a9f29f8.